### PR TITLE
[#5050] Changed logic to permit VLANs assigned to a device's location's parent locations (including parents of parents, etc.) to be assigned to that device's interfaces.

### PR DESCRIPTION
**Pull request by HanlinMiao**
Re-opened PR for https://github.com/nautobot/nautobot/pull/6091 with passing unittests

### CODE REVIEW

File: nautobot/dcim/api/serializers.py [hunk 1]
```diff
@@ -798,13 +798,19 @@ class Meta:
...
         # Validate many-to-many VLAN assignments
         device = self.instance.device if self.instance else data.get("device")
-        # TODO: after Location model replaced Site, which was not a hierarchical model, should we allow users to assign a VLAN belongs to
-        # the parent Location or the child location of `device.location`?
+        location = None
+        if device:
+            location = device.location
+        if location:
+            location_ids = [ancestor.id for ancestor in location.ancestors()]
+            location_ids.append(location.id)
```
Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:806]: Possibly:

```suggestion
            location_ids = location.ancestors(include_self=True).values_list("id", flat=True)
```

File: nautobot/dcim/api/serializers.py [hunk 2]
```diff
@@ -798,13 +798,19 @@ class Meta:
...
+        else:
+            location_ids = []
         for vlan in data.get("tagged_vlans", []):
-            if vlan.locations.exists() and not vlan.locations.filter(pk=device.location.pk).exists():
+            if vlan.locations.exists() and not vlan.locations.filter(pk__in=location_ids).exists():
                 raise serializers.ValidationError(
                     {
-                        "tagged_vlans": f"VLAN {vlan} must have a common location as the interface's parent device, or "
+                        "tagged_vlans": f"VLAN {vlan} must have the same location as the interface's parent device, or is in one of the parents of the interface's parent device's location, or "
```
Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:813]: Mind splitting this string across lines to keep it to 120 characters?

File: nautobot/dcim/forms.py [hunk 3]
```diff
@@ -213,23 +213,28 @@ def clean(self):
...
-        # TODO: after Location model replaced Site, which was not a hierarchical model, should we allow users to add a VLAN
-        # belongs to the parent Location or the child location of the parent device to the `tagged_vlan` field of the interface?
+        # Validate tagged VLANs; must be a global VLAN or in the same location as the
+        # parent device/VM or any of that location's parent locations
         elif mode == InterfaceModeChoices.MODE_TAGGED:
-            valid_location = self.cleaned_data[parent_field].location
+            location = self.cleaned_data[parent_field].location
+            if location:
+                location_ids = [ancestor.id for ancestor in location.ancestors()]
+                location_ids.append(location.id)
```
Comment (glennmatthews) [inline nautobot/dcim/forms.py:222]:
```suggestion
                location_ids = location.ancestors(include_self=True).values_list("id", flat=True)
```

File: nautobot/dcim/tests/test_forms.py [hunk 4]
```diff
@@ -343,9 +343,20 @@ def setUpTestData(cls):
 
     def test_interface_form_clean_vlan_location_fail(self):
         """Assert that form validation fails when no matching locations are associated to tagged VLAN"""
-        self.vlan.locations.set(list(Location.objects.exclude(pk=self.device.location.pk))[:2])
+        location = self.device.location
+        location_ids = [ancestor.id for ancestor in location.ancestors()]
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_forms.py:347]: Do we need an `include_self=True` here?

File: nautobot/dcim/tests/test_forms.py [hunk 5]
```diff
@@ -343,9 +343,20 @@ def setUpTestData(cls):
...
         form = InterfaceForm(data=self.data, instance=self.interface)
         self.assertFalse(form.is_valid())
+        self.vlan.locations.set([location.id])
+        self.data["tagged_vlans"] = [self.vlan]
+        form = InterfaceForm(data=self.data, instance=self.interface)
+        self.assertTrue(form.is_valid())
+        self.vlan.locations.set(location_ids[:2])
+        self.data["tagged_vlans"] = [self.vlan]
+        form = InterfaceForm(data=self.data, instance=self.interface)
+        self.assertTrue(form.is_valid())
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_forms.py:359]: Noting that the test function is called `form_clean_vlan_location_fail`, should these success cases be a separate test?

File: nautobot/dcim/utils.py [hunk 6]
```diff
@@ -14,7 +14,7 @@
     SCRAPLI_LIB_MAPPER_REVERSE,
 )
 
-from nautobot.core.choices import ColorChoices
+from nautobot.core.utils.color import hex_to_rgb, lighten_color, rgb_to_hex
```
Comment (glennmatthews) [inline nautobot/dcim/utils.py:17]: bad merge?

File: nautobot/dcim/utils.py [hunk 7]
```diff
@@ -55,14 +55,11 @@ def cable_status_color_css(record):
...
-            ColorChoices.COLOR_AMBER: "warning",
-            ColorChoices.COLOR_CYAN: "info",
-        }
-        status_color = record.cable.get_status_color().strip("#")
-        return CABLE_STATUS_TO_CSS_CLASS.get(status_color, "")
+    # The status colors are for use with labels and such, and tend to be quite bright.
+    # For this function we want a much milder, mellower color suitable as a row background.
+    base_color = record.cable.get_status_color().strip("#")
+    lighter_color = rgb_to_hex(*lighten_color(*hex_to_rgb(base_color), 0.75))
+    return f"background-color: #{lighter_color}"
```
Comment (glennmatthews) [inline nautobot/dcim/utils.py:62]: bad merge?

File: nautobot/dcim/utils.py [hunk 8]
```diff
@@ -139,20 +136,23 @@ def validate_interface_tagged_vlans(instance, model, pk_set):
...
     # Filter the model objects based on the primary keys passed in kwargs and exclude the ones that have
-    # a location that is not the parent's location or None
-    # TODO: after Location model replaced Site, which was not a hierarchical model, should we allow users to add a VLAN
-    # belongs to the parent Location or the child location of the parent device to the `tagged_vlan` field of the interface?
-    device_location = getattr(instance.parent, "location", None)
+    # a location that is not the parent's location, or parent's location's ancestors, or None
+    location = getattr(instance.parent, "location", None)
+    if location:
+        location_ids = [ancestor.id for ancestor in location.ancestors()]
+        location_ids.append(location.id)
```
Comment (glennmatthews) [inline nautobot/dcim/utils.py:143]:
```suggestion
        location_ids = location.ancestors(include_self=True).values_list("id", flat=True)
```

File: nautobot/dcim/utils.py [hunk 9]
```diff
@@ -139,20 +136,23 @@ def validate_interface_tagged_vlans(instance, model, pk_set):
...
+        model.objects.filter(pk__in=pk_set).exclude(locations__isnull=True).exclude(locations__in=location_ids)
     )
 
     if tagged_vlans.count():
         raise ValidationError(
             {
                 "tagged_vlans": (
                     f"Tagged VLAN with names {list(tagged_vlans.values_list('name', flat=True))} must all belong to the "
-                    f"same location as the interface's parent device, or it must be global."
+                    f"same location as the interface's parent device, one of the parent locations of the interface's parent device's location, or it must be global."
```
Comment (glennmatthews) [inline nautobot/dcim/utils.py:155]: split string to 120 characters please

File: nautobot/dcim/tests/test_forms.py (hunk 4 above)
Comment (HanlinMiao) [reply nautobot/dcim/tests/test_forms.py:347]: Yeah I believe so thanks for catching that

**Pull request merged**